src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs

```diff
@@ -666,6 +666,44 @@ public static bool HaveSameConstraints(TypeParameterSymbol typeParameter1, TypeM
                 AreConstraintTypesSubset(substitutedTypes2, substitutedTypes1, typeParameter1);
         }
 
+        public static bool HaveSameNullabilityInConstraints(TypeParameterSymbol typeParameter1, TypeMap typeMap1, TypeParameterSymbol typeParameter2, TypeMap typeMap2)
+        {
+            if (!typeParameter1.IsValueType)
+            {
+
+                bool? isNotNullableIfReferenceType1 = typeParameter1.IsNotNullableIfReferenceType;
+                bool? isNotNullableIfReferenceType2 = typeParameter2.IsNotNullableIfReferenceType;
+                if (isNotNullableIfReferenceType1.HasValue && isNotNullableIfReferenceType2.HasValue && 
+                    isNotNullableIfReferenceType1.GetValueOrDefault() != isNotNullableIfReferenceType2.GetValueOrDefault())
+                {
+                    return false;
+                }
+            }
+
+            // Check that constraintTypes1 is a subset of constraintTypes2 and
+            // also that constraintTypes2 is a subset of constraintTypes1
+
+            var constraintTypes1 = typeParameter1.ConstraintTypesNoUseSiteDiagnostics;
+            var constraintTypes2 = typeParameter2.ConstraintTypesNoUseSiteDiagnostics;
+
+            // The two sets of constraints may differ in size but still be considered
+            // the same (duplicated constraints, ignored "object" constraints), but
+            // if both are zero size, the sets must be equal.
+            if ((constraintTypes1.Length == 0) && (constraintTypes2.Length == 0))
+            {
+                return true;
+            }
+
+            var substitutedTypes1 = new HashSet<TypeSymbol>(TypeSymbol.EqualsAllIgnoreOptionsPlusNullablWitUnknownMatchesAnyComparer);
+            var substitutedTypes2 = new HashSet<TypeSymbol>(TypeSymbol.EqualsAllIgnoreOptionsPlusNullablWitUnknownMatchesAnyComparer);
+
+            SubstituteConstraintTypes(constraintTypes1, typeMap1, substitutedTypes1);
+            SubstituteConstraintTypes(constraintTypes2, typeMap2, substitutedTypes2);
+
+            return AreConstraintTypesSubset(substitutedTypes1, substitutedTypes2, typeParameter2) &&
+                AreConstraintTypesSubset(substitutedTypes2, substitutedTypes1, typeParameter1);
```

Comment:
Consider extracting a helper method (with comparer parameter) for use here and in `HaveSameConstraints`.